pkg/hub/managedclusterset/defaultclustersetcontroller.go

```diff
@@ -0,0 +1,118 @@
+package managedclusterset
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"time"
+
+	"github.com/openshift/library-go/pkg/controller/factory"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	clustersetv1beta1 "open-cluster-management.io/api/client/cluster/clientset/versioned/typed/cluster/v1beta1"
+	clusterinformerv1beta1 "open-cluster-management.io/api/client/cluster/informers/externalversions/cluster/v1beta1"
+	clusterlisterv1beta1 "open-cluster-management.io/api/client/cluster/listers/cluster/v1beta1"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+const defaultManagedClusterSetName = "defaultmanagedclusterset"
+
+// TODO: set the default value of spec
+var defaultManagedClusterSetSpec = clusterv1beta1.ManagedClusterSetSpec{}
+
+type defaultManagedClusterSetController struct {
+	clusterSetClient clustersetv1beta1.ClusterV1beta1Interface
+	clusterSetLister clusterlisterv1beta1.ManagedClusterSetLister
+	eventRecorder    events.Recorder
+}
+
+func NewDefaultManagedClusterSetController(
+	clusterSetClient clustersetv1beta1.ClusterV1beta1Interface,
+	clusterSetInformer clusterinformerv1beta1.ManagedClusterSetInformer,
+	recorder events.Recorder) factory.Controller {
+
+	c := &defaultManagedClusterSetController{
+		clusterSetClient: clusterSetClient,
+		clusterSetLister: clusterSetInformer.Lister(),
+		eventRecorder:    recorder.WithComponentSuffix("default-managed-cluster-set-controller"),
+	}
+
+	return factory.New().
+		WithFilteredEventsInformersQueueKeyFunc(
+			func(obj runtime.Object) string {
+				accessor, _ := meta.Accessor(obj)
+				return accessor.GetName()
+			},
+			func(obj interface{}) bool {
+				metaObj, ok := obj.(metav1.ObjectMetaAccessor)
+				if !ok {
+					return false
+				}
+				// filter clustersets except defaultManagedClusterSet.
+				return defaultManagedClusterSetName != metaObj.GetObjectMeta().GetName()
+			},
+			clusterSetInformer.Informer(),
+		).
+		WithSync(c.sync).
+		// set the resync period to 10s.
+		ResyncEvery(10*time.Second).
```

Comment:
@elgnay do we need this?

Reply:
> @elgnay do we need this?

`ResyncEvery` is invoked to make sure the default clusterset will be recreated once it is deleted for some reason. If we have a validation webhook it prevents from deleting the default clusterset, this logic is not necessary any more.